### PR TITLE
QE: Add Cobbler buildiso tests

### PR DESCRIPTION
**Pull request by nodeg**
## What does this PR change?

This PR will

- extend the already implemented Cobbler steps to include checking for profiles
- add tests for Cobbler `buildiso`

in preparation of including Cobbler 3.3.2 to Manager 4.3 Beta 3.

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
- Cucumber tests were added

- [x] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/17250

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: testsuite/features/step_definitions/common_steps.rb
Comment (nodeg) [reply testsuite/features/step_definitions/common_steps.rb:1395]: It seems we disabled this particular check in RuboCop.

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (srbarrios) [inline testsuite/features/secondary/srv_cobbler_buildiso.feature:32]: Same, not sure if we really need to do this check in several scenario.

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (srbarrios) [inline testsuite/features/secondary/srv_cobbler_buildiso.feature:11]: I wonder if we should make this step more specific for cobbler, haven't we other mechanism for autoinstall?

File: testsuite/features/support/cobbler_test.rb
Comment (srbarrios) [inline testsuite/features/support/cobbler_test.rb:58]: I think it's not a good practice to raise an exception in a rescue section.
What about just `puts` this error?

File: testsuite/features/support/cobbler_test.rb
Comment (srbarrios) [inline testsuite/features/support/cobbler_test.rb:74]: OK, apparently we are raising exception in the rescue section in that whole class :)
So, then maybe that's a refactor for another moment... but I would consider refactor and not raise an exception inside a rescue.

File: testsuite/features/step_definitions/common_steps.rb
Comment (srbarrios) [inline testsuite/features/step_definitions/common_steps.rb:1447]: For item, instead of using `item "([^"]*)"`, we might want to use `item "(distro|profile|system)"`

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (vandabarata) [inline testsuite/features/secondary/srv_cobbler_buildiso.feature:21]: I would substitute these redundant `Then`s with `And`s, after the first `Then` so it's easier to read and more consistent with the rest of the testsuite :thinking:

File: testsuite/features/step_definitions/common_steps.rb
Comment (srbarrios) [reply testsuite/features/step_definitions/common_steps.rb:1447]: And I would personally avoid the whole comment, as it's kind of obvious that we will need to check Cobbler docs or man page for more info.

File: testsuite/features/step_definitions/common_steps.rb
Comment (srbarrios) [inline testsuite/features/step_definitions/common_steps.rb:1454]: If only some parameters are available, maybe we want to use a regular expression with ORs.

Comment (srbarrios) [reply testsuite/features/step_definitions/common_steps.rb:1454]: We can also avoid the comment

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (vandabarata) [inline testsuite/features/secondary/srv_cobbler_buildiso.feature:35]: Same as before, this would be an `And` for me

File: testsuite/features/step_definitions/common_steps.rb
Comment (srbarrios) [inline testsuite/features/step_definitions/common_steps.rb:1439]: This line looks too long for me... we should make it shorter.

File: testsuite/features/support/cobbler_test.rb
Comment (vandabarata) [inline testsuite/features/support/cobbler_test.rb:123]: could this not be a single function, getting an argument of distro, profile or system? it seems to be the only thing different between them :thinking:

File: testsuite/features/step_definitions/common_steps.rb
Comment (srbarrios) [inline testsuite/features/step_definitions/common_steps.rb:1435]: Looks good like that, but just in case you want to use it for other cases, we can also have stored files in upload_files folder inside our testsuite folder. Then we upload the file to the server at the path we want.

File: testsuite/features/step_definitions/common_steps.rb
Comment (vandabarata) [inline testsuite/features/step_definitions/common_steps.rb:1433]: I know this still works like this, but I believe we're always using a `Given`, `When` or `Then` when defining the steps themselves so I'd rearrange this accordingly

**Comment by srbarrios** [comment]
One last suggestion, it might be interesting to start organizing some step definitions in different ruby files.
What about moving all autoinstall/cobbler steps to a new `autoinstall_steps.rb` file?

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (nodeg) [reply testsuite/features/secondary/srv_cobbler_buildiso.feature:11]: We already had it in the testsuite, so I copied it from https://github.com/uyuni-project/uyuni/blob/57ae43ccf9750f42190cfcd160f7d6251b39d4a5/testsuite/features/secondary/srv_distro_cobbler.feature#L10-L11

File: testsuite/features/support/cobbler_test.rb
Comment (nodeg) [reply testsuite/features/support/cobbler_test.rb:58]: We can use `puts` here of course. I just stuck to the same style the other methods in this class already used. Nearly all of them use `raise` inside a `rescue`.

File: testsuite/features/support/cobbler_test.rb
Comment (nodeg) [reply testsuite/features/support/cobbler_test.rb:74]: Agreed. I will open another issue in our board about improving this.

File: testsuite/features/step_definitions/common_steps.rb
Comment (nodeg) [reply testsuite/features/step_definitions/common_steps.rb:1454]: I would prefer looking those parameters up in the docs/code since there are several ones. This makes us more independent and we do not have to touch our code if there are changes upstream.

File: testsuite/features/support/cobbler_test.rb
Comment (nodeg) [reply testsuite/features/support/cobbler_test.rb:123]: Of course we can do this :wink:  As mentioned earlier to Oscar, I just extended the code and stuck to the style already present. I will mention this is the issue I create for improving/refactoring the `CobblerTest` class.

File: testsuite/features/step_definitions/common_steps.rb
Comment (nodeg) [reply testsuite/features/step_definitions/common_steps.rb:1439]: I copied this over from the upstream integration test:
https://github.com/cobbler/cobbler/blob/f440affa5cb9d41e589033be5c3511f18fcac348/system-tests/tests/buildiso-%40#L59-L60. I did not found a better solution to shorten this, yet.

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (nodeg) [reply testsuite/features/secondary/srv_cobbler_buildiso.feature:18]: I did include it in most steps since Cobbler could possibly fail in between 2 steps but I am not sure how often this could happen. I removed all of them and left just the first one.

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (nodeg) [reply testsuite/features/secondary/srv_cobbler_buildiso.feature:32]: Fixed.

File: testsuite/features/step_definitions/common_steps.rb
Comment (nodeg) [reply testsuite/features/step_definitions/common_steps.rb:1447]: Fixed.

File: testsuite/features/step_definitions/common_steps.rb
Comment (nodeg) [reply testsuite/features/step_definitions/common_steps.rb:1433]: Fixed.

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (nodeg) [reply testsuite/features/secondary/srv_cobbler_buildiso.feature:35]: Fixed.

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (nodeg) [reply testsuite/features/secondary/srv_cobbler_buildiso.feature:21]: Fixed.

File: testsuite/features/support/cobbler_test.rb
Comment (nodeg) [reply testsuite/features/support/cobbler_test.rb:74]: https://github.com/SUSE/spacewalk/issues/17382

File: testsuite/features/support/cobbler_test.rb
Comment (nodeg) [reply testsuite/features/support/cobbler_test.rb:123]: https://github.com/SUSE/spacewalk/issues/17382

**Comment by vandabarata** [comment]
LGTM, this feels like a behemoth of a feature, and my comments were mainly nitpicks - great job here :grin: only up to @srbarrios to give the go ahead now :stuck_out_tongue:

**Comment by nodeg** [comment]
>  What about moving all autoinstall/cobbler steps to a new `autoinstall_steps.rb` file?

Good idea. Since Cobbler handles the autoinstallation and other things I would prefer a `cobbler_steps.rb` file.

**Review by vandabarata — APPROVED** [review]

File: testsuite/features/secondary/srv_cobbler_buildiso.feature
Comment (srbarrios) [reply testsuite/features/secondary/srv_cobbler_buildiso.feature:12]: OK, then fine to don't change it for this PR.

File: testsuite/features/step_definitions/common_steps.rb
Comment (srbarrios) [reply testsuite/features/step_definitions/common_steps.rb:1439]: Yup, that's what I was suggesting :)
I guess Rubocop don't complains just because we are disabling the Cop about it :)

File: testsuite/features/support/cobbler_test.rb
Comment (srbarrios) [reply testsuite/features/support/cobbler_test.rb:58]: Ok to keep the same rescue management on this PR.

File: testsuite/features/step_definitions/common_steps.rb
Comment (nodeg) [reply testsuite/features/step_definitions/common_steps.rb:1439]: Ok. I will adjust this with another commit.

File: testsuite/features/step_definitions/common_steps.rb
Comment (srbarrios) [reply testsuite/features/step_definitions/common_steps.rb:1451]: Keep in mind that BDD and Gherkin documentation is designed to have living documentation.
So the tests in Cucumber serves to several purposes, and one is to be attached and dependant to product code in the sense that it reflects exactly what the product code offers.

So, exposing here the available parameters, is a good strategy.
Anyway, if you prefer to keep it like that, fine for me.

**Review by srbarrios — APPROVED** [review]

**Comment by nodeg** [comment]
Thanks for approving. I am not sure when to merge this, since it is for Manager 4.3 Beta 3 and I am not sure if the Cobbler package in the different places is ready, yet. Do you have more information on that @SchoolGuy @meaksh?

ATM I am using a special test sumaform environment for writing those tests:

```hcl
  server = {
    additional_repos = {
      server_stack = "http://download.suse.de/ibs/Devel:/Galaxy:/Manager:/TEST/images/repo/SLE-Module-SUSE-Manager-Server-4.3-POOL-x86_64-Media1/"
      Test_repo = "http://download.suse.de/ibs/home:/PSuarezHernandez:/branches:/openSUSE.org:/systemsmanagement/SLE_15_SP4/"
    }
    additional_packages = [ "cobbler" ]
  }
```

**Comment by SchoolGuy** [comment]
Yeah. Basically once we have merged https://github.com/uyuni-project/uyuni/pull/4977 we can rebase this and then merge it. At least if I havn't forgotten anything.

**Comment by nodeg** [comment]
Ok, so we are waiting for https://github.com/uyuni-project/uyuni/pull/4977. Thank you Enno!

**Comment by srbarrios** [comment]
> Ok, so we are waiting for #4977. Thank you Enno!

@SchoolGuy @nodeg what we do in other squads is to merge the tests of the feature in the feature branch, so in that case this PR will merge inside [master-fixes-for-new-cobbler](https://github.com/uyuni-project/uyuni/tree/master-fixes-for-new-cobbler) and so from QE we don't need to worry about when the feature branch is merged, that it will go all in the same pack.

**Comment by juliogonzalez** [comment]
What's this mess? :-?

I see releng was added, but to me it looks like this PR was incorrectly rebased or something?

**Comment by nodeg** [comment]
Sorry @juliogonzalez. @srbarrios and I tried to change the destination branch but it did not work the way it should be. I open another PR and close this one.
Closed in favor of #5142.

**Pull request closed without merging**
